x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview/engine_overview_logic.ts

```diff
@@ -0,0 +1,130 @@
+import { HttpLogic } from '../../../shared/http';
+import { EngineLogic } from '../engine';
+
+const POLLING_DURATION = 5000;
+
+interface EngineOverviewApiData {
+  apiLogsUnavailable: boolean;
+  documentCount: number;
+  startDate: string;
+  endDate: string;
+  operationsPerDay: number[];
+  queriesPerDay: number[];
+  totalClicks: number;
+  totalQueries: number;
+}
+interface EngineOverviewValues extends EngineOverviewApiData {
+  dataLoading: boolean;
+  timeoutId: number | null;
+}
+
+interface EngineOverviewActions {
+  setPolledData(engineMetrics: EngineOverviewApiData): EngineOverviewApiData;
+  setTimeoutId(timeoutId: number): { timeoutId: number };
+  pollForOverviewMetrics(): void;
+  onPollingSuccess(engineMetrics: EngineOverviewApiData): EngineOverviewApiData;
+}
+
+export const EngineOverviewLogic = kea<MakeLogicType<EngineOverviewValues, EngineOverviewActions>>({
+  path: ['enterprise_search', 'app_search', 'engine_overview_logic'],
+  actions: () => ({
+    setPolledData: (engineMetrics) => engineMetrics,
+    setTimeoutId: (timeoutId) => ({ timeoutId }),
+    pollForOverviewMetrics: true,
+    onPollingSuccess: (engineMetrics) => engineMetrics,
+  }),
+  reducers: () => ({
+    dataLoading: [
+      true,
+      {
+        setPolledData: () => false,
+      },
+    ],
+    apiLogsUnavailable: [
+      false,
+      {
+        setPolledData: (_, { apiLogsUnavailable }) => apiLogsUnavailable,
+      },
+    ],
+    startDate: [
+      '',
+      {
+        setPolledData: (_, { startDate }) => startDate,
+      },
+    ],
+    endDate: [
+      '',
+      {
+        setPolledData: (_, { endDate }) => endDate,
+      },
+    ],
+    queriesPerDay: [
+      [],
+      {
+        setPolledData: (_, { queriesPerDay }) => queriesPerDay,
+      },
+    ],
+    operationsPerDay: [
+      [],
+      {
+        setPolledData: (_, { operationsPerDay }) => operationsPerDay,
+      },
+    ],
+    totalQueries: [
+      0,
+      {
+        setPolledData: (_, { totalQueries }) => totalQueries,
+      },
+    ],
+    totalClicks: [
+      0,
+      {
+        setPolledData: (_, { totalClicks }) => totalClicks,
+      },
+    ],
+    documentCount: [
+      0,
+      {
+        setPolledData: (_, { documentCount }) => documentCount,
+      },
+    ],
+    timeoutId: [
+      null,
+      {
+        setTimeoutId: (_, { timeoutId }) => timeoutId,
+      },
+    ],
+  }),
+  listeners: ({ actions }) => ({
+    pollForOverviewMetrics: async () => {
+      const { http } = HttpLogic.values;
+      const { engineName } = EngineLogic.values;
+
+      try {
+        const response = await http.get(`/api/app_search/engines/${engineName}/overview`);
+        actions.onPollingSuccess(response);
+      } catch (e) {
+        flashAPIErrors(e);
+      }
+    },
+    onPollingSuccess: (engineMetrics) => {
+      const timeoutId = window.setTimeout(actions.pollForOverviewMetrics, POLLING_DURATION);
+      actions.setTimeoutId(timeoutId);
```

Comment:
Don't chat this, this is non-blocking feedback. Just wanted to point out a best practice from Kea and Redux.

https://kea.js.org/docs/guide/concepts#listeners

> If you read the reducers section above, you'll remember that it's an anti-pattern to only have setThis and setThat actions that only update this or that.

Often times when you need to call 2 or more actions like this on success to set resulting data, it's a good indication that you're following this anti-pattern.

There really is only 1 "action" happening here. Meaning, if you think of actions as "events" like Kea suggestion, the "event" that is happening here is that polling has completed. Reducers should react to that "event" and set data accordingly.

I could be mis-reading this, just thought I'd take a moment to point this out.

Reply:
Ahh super interesting! I went back and forth a bit on whether I should refactor/clean this up as well actually but ended up leaving it as-is because it didn't bother me too much.

>  If you read the reducers section above, you'll remember that it's an anti-pattern to only have setThis and setThat actions that only update this or that.

This confuses me a bit, we do this in a number of our Kea files and I'm not quite clear on what the "correct" code looks like (maybe just a bit too early in the morning for me).

I did just discover [sharedListeners](https://kea.js.org/docs/guide/advanced/#shared-listeners) in their docs which blew my mind (how long has that been there!?) and looks like we could use it in our code... I'd love to do an overall Kea refactor at some point if we have time, if we have it 😅